activestorage/app/models/active_storage/blob.rb

```diff
@@ -349,6 +349,7 @@ def compute_checksum_in_chunks(io)
       raise ArgumentError, "io must be rewindable" unless io.respond_to?(:rewind)
 
       OpenSSL::Digest::MD5.new.tap do |checksum|
+        io.rewind
```

Comment:
I think we should save `IO#pos`, and reset to that offset. Otherwise it will also fail is you pass an IO with a non-zero offset.

Reply:
Thank you for taking a look!

I think it works for partially-consumed streams but the entire stream is uploaded, which is maybe not what one would expect. I added a test to demonstrate this (5b3f41a).

I also tried your suggestion of preserving the offset (e76bc70). It seems to work with the disk service but I'm not sure it will for others. For instance, it seems the Ruby client library for Google Cloud Storage rewinds the stream before the upload, so we'd end up getting an integrity error.

Thus, I'm wondering if we wouldn't be better off either always rewinding or bailing out early with an helpful error for partially-consumed streams. Do you have a steer?

Reply:
>  For instance, it seems the Ruby client library for Google Cloud Storage rewinds the stream before the upload, so we'd end up getting an integrity error.

Argh, that's really unfortunate. I went to check the code, and indeed, it doesn't care about offsets.

> Do you have a steer?

I'm not the most familiar with Active Storage, but we probably should try to be consistent. So you initial rewinf is probably the way to go, but my worry is some people might have been relying on that behavior on some backends.

So we should add a CHANGELOG etc.

Reply:
To clarify, I think the behavior is currently consistent across all backends: when given a stream with an offset, Active Storage raises an `IntegrityError` (adb6f69).

That's because Active Storage rewinds the stream before handing it over to the backend, namely right after the checksum is calculated: https://github.com/rails/rails/blob/v8.0.0/activestorage/app/models/active_storage/blob.rb#L357

So rewinding the stream shouldn't be breaking, I think? Although I appreciate that the resulting behavior (i.e. uploading the whole stream) might be surprising if the stream has an offset and is not at EOF. In that case, maybe we're better off raising an error. Do you have any thoughts on  d25676c?

Reply:
> Do you have any thoughts on [d25676c](https://github.com/rails/rails/commit/d25676c535c6fb11c2ec61e9f42fe034c1f501dc)?

I think it's a good idea. But I'm not a big user of Active Storage so no strong opinion.

Reply:
OK, thank you again for your time and feedback!

Reply:
@jonathanhefner, I hope you don’t mind the mention. I noticed you contributed to Active Storage recently-ish and thought you might be well-positioned to have an opinion on this pull request. I’d appreciate your thoughts if you have the time. Thank you!